File: src/ghsingo/internal/audio/bed.go
package audio

import "math/rand"

// BedVoice is a long-lived noise pad that supplies the L1 mid-layer
// texture. Pink-ish noise is run through a single-pole lowpass to keep
// it under the drone's brightness, then amplitude-smoothed toward a
// brightness-driven target.
type BedVoice struct {
	sampleRate int
	rng        *rand.Rand

	// Voss-McCartney style pink noise generators (cheap pink approx).
	rows    [5]float32
	counter int

	lp1 float32
	lp2 float32
	a   float32 // lowpass coefficient

	targetGain float32
	gain       float32
	gainSlew   float32
}

// NewBedVoice constructs a bed voice with cutoff ~600 Hz so the texture
// never competes with the accent layer above 1 kHz.
func NewBedVoice(sampleRate int, seed int64) *BedVoice {
	if seed == 0 {
		seed = 0x6361 // arbitrary non-zero default — silent renders are
		// not what we want when callers forget the seed
	}
	return &BedVoice{
		sampleRate: sampleRate,
		rng:        rand.New(rand.NewSource(seed)),
		a:          0.04, // ~600 Hz at 44.1 kHz
		gainSlew:   1.0 / float32(sampleRate) * 3.0,
	}
}

// SetTargetGain sets the smoothed amplitude target.
func (b *BedVoice) SetTargetGain(g float32) {
	if g < 0 {
		g = 0
	}
	if g > 1 {
		g = 1
	}
	b.targetGain = g
}

// NextSample produces one mono pink-noise sample, lowpassed, scaled by
// the smoothed gain.
func (b *BedVoice) NextSample() float32 {
	b.counter++
	for i := 0; i < 5; i++ {
		if b.counter%(1<<i) == 0 {
			b.rows[i] = b.rng.Float32()*2 - 1
		}
	}
	var pink float32
	for _, r := range b.rows {
		pink += r
	}
	pink *= 0.2

	b.lp1 += b.a * (pink - b.lp1)
	b.lp2 += b.a * (b.lp1 - b.lp2)

	if b.gain < b.targetGain {
		b.gain += b.gainSlew
		if b.gain > b.targetGain {
			b.gain = b.targetGain
		}
	} else if b.gain > b.targetGain {
		b.gain -= b.gainSlew
		if b.gain < b.targetGain {
			b.gain = b.targetGain
		}
	}

	return b.lp2 * b.gain
}

File: src/rsghsing/src/audio/bed.rs
//! L1 pink-noise bed. Ported from `internal/audio/bed.go`.
//!
//! Go seeds the Voss-McCartney rows from `rand.NewSource(0xb1d)` — a fixed
//! seed — so this layer is bit-deterministic on both sides.

use crate::gorand::GoRand;

const BED_SEED: i64 = 0xb1d;

pub struct BedVoice {
    sample_rate: i32,
    rng: GoRand,
    rows: [f32; 5],
    counter: i64,
    lp1: f32,
    lp2: f32,
    a: f32,
    target_gain: f32,
    gain: f32,
    gain_slew: f32,
}

impl BedVoice {
    pub fn new(sample_rate: i32) -> Self {
        BedVoice {
            sample_rate,
            rng: GoRand::new(BED_SEED),
            rows: [0.0; 5],
            counter: 0,
            lp1: 0.0,
            lp2: 0.0,
            a: 0.04, // ~600 Hz at 44.1 kHz
            target_gain: 0.0,
            gain: 0.0,
            gain_slew: 1.0 / sample_rate as f32 * 3.0,
        }
    }

    pub fn set_target_gain(&mut self, g: f32) {
        self.target_gain = g.clamp(0.0, 1.0);
    }

    pub fn next_sample(&mut self) -> f32 {
        self.counter += 1;
        for i in 0..5i32 {
            if self.counter % (1i64 << i) == 0 {
                self.rows[i as usize] = (self.rng.float64() * 2.0 - 1.0) as f32;
            }
        }
        let mut pink = 0.0f32;
        for r in self.rows {
            pink += r;
        }
        pink *= 0.2;

        self.lp1 += self.a * (pink - self.lp1);
        self.lp2 += self.a * (self.lp1 - self.lp2);

        if self.gain < self.target_gain {
            self.gain += self.gain_slew;
            if self.gain > self.target_gain {
                self.gain = self.target_gain;
            }
        } else if self.gain > self.target_gain {
            self.gain -= self.gain_slew;
            if self.gain < self.target_gain {
                self.gain = self.target_gain;
            }
        }

        self.lp2 * self.gain
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bed_is_deterministic_for_a_given_seed() {
        let run = || {
            let mut b = BedVoice::new(44_100);
            b.set_target_gain(0.2);
            (0..5_000).map(|_| b.next_sample()).collect::<Vec<_>>()
        };
        assert_eq!(run(), run());
    }

    #[test]
    fn lowpass_keeps_output_bounded() {
        let mut b = BedVoice::new(44_100);
        b.set_target_gain(1.0);
        for _ in 0..44_100 {
            b.next_sample();
        }
        let peak = (0..44_100).map(|_| b.next_sample()).fold(0.0f32, f32::max);
        assert!(peak <= 1.0, "peak {peak}");
    }
}
